### ISSUE TITLE
Autocomplete input shows wrong selected options

### ISSUE THREAD

**Issue by panfiva** [marmelab/react-admin#9490]
I have an AutocompleteArrayInput with the following options
```js
const options = [
  { label: 'movies', id: 1, count: 1 },
  { label: 'games', id: 2, count: 1 },
]
```
Option display names are computed using `${label} (${count}` attribute

When options are updated to change the count, currently-selected options still show old counts. Note - when reproducing the issue, I accidently selected react-admin v3 as template, and the issue was not present there. It is present in the current version of react-admin, though

**What you were expecting:**
Labels for currently-selected values should be updated to include new counts

**Steps to reproduce:**
- Go to https://stackblitz.com/edit/github-yuwnpr?file=src%2Fposts%2FPostEdit.tsx
- Navigate to "Posts" and click any record to edit
- Select a value from the drop down
- click `increase count` button
- observe that selected label is still the same, while drop-down options were updated

![image](https://github.com/marmelab/react-admin/assets/48455289/4f6e3a45-515a-4f14-83e8-7d3607d4fb3e)

**Related code:**

```ts
const PostEdit = () => {
  const [options, setOptions] = React.useState([
    { label: 'movies', id: 1, count: 1 },
    { label: 'games', id: 2, count: 1 },
  ]);

  const onClick = () => {
    const newOptions = options.map((v) => ({ ...v, count: v.count + 1 }));
    setOptions(newOptions);
  };

  const optionText = (choice: any) => `${choice.label} (${choice.count})`;

  const isOptionEqualToValue = (option: any, value: any) => {
    return option.label === value.label;
  };

  const renderTags = (values: Record<string, any>, getTagProps: any) => {
    return values.map((option: any, index: number) => {
      const name = optionText(option);
      return (
        <Tooltip key={option.id} title={name} placement="right-end">
          <Chip
            key={option.id}
            label={name}
            size="small"
            {...getTagProps({ index })}
          />
        </Tooltip>
      );
    });
  };

  return (
    <Edit>
      <TabbedForm>
        <FormTab label="post.form.summary 4">
          <AutocompleteArrayInput
            disableCloseOnSelect
            blurOnSelect={false}
            optionText={optionText}
            optionValue="id"
            isOptionEqualToValue={isOptionEqualToValue}
            label="label2"
            multiple
            choices={options}
            isLoading={false}
            renderTags={renderTags}
            source="test"
          />

          <Button sx={{ mt: 5 }} onClick={onClick}>
            Increase count
          </Button>
          {JSON.stringify(options)}
        </FormTab>
      </TabbedForm>
    </Edit>
  );
};
```

**Other information:**

**Environment**

* React-admin version:
* Last version that did not exhibit the issue (if applicable):
* React version:
* Browser:
* Stack trace (in case of a JS error):

**Comment by slax57** [comment]
Thanks for the detailed report and the stackblitz.

I have been able to reproduce the issue. 
However, I'm not sure how to qualify it.

Indeed, I can't think of any practical use case where you would need to update the label of a selected value _after_ it has been selected.

Can you tell us more about your use case and why you need to update the label after selection?

It's also worth it to note the label correctly updates as soon as the value changes (i.e. when another option is selected).

Thanks.

**Comment by panfiva** [comment]
The use case as follows:
We have a `projects` resource that has different filters available (e.g. `start_date`,  `project _type`, `isActive`).

Project_type filter is an `AutocompleteArrayInput` with options computed using `select project_type, count(*) as count from projects where <...other filters> group by project_type` query

We want labels to indicate project count to make it easier for users to understand the environment.

For example:

1) Start with no filters: Autocomplete shows:
 - project A (35)
 - project B (10)
 - project C (10)

2) Select filter `project_type=project A`
- project A (35) - Autocomplete selected shows correct value
- project B (10) - Autocomplete drop-down shows correct value
- project C (10)  - Autocomplete drop-down shows correct value

 2) Apply `start_date>=2023-01-01` (assume this filter will reduce project count by 5)
- project A (35) - Autocomplete selected value is incorrect; we want users to see new count
- project B (5) - Autocomplete drop-down shows correct value
- project C (5)  - Autocomplete drop-down shows correct value

**Comment by slax57** [comment]
Thanks for this additional information.
I guess it makes sense indeed.
I'll label this issue as a bug.

Unfortunately, I have spent some time trying to find a possible root cause (like a missing dependency in an effect or a memo), but with no luck. We also can't rule out for now that this issue may exist in MUI's Autocomplete.
TLDR, this will require more investigation work.

Thanks for the report anyway!

**Comment by slax57** [comment]
@panfiva 

I don't know why I can't manage to open the codesandbox you provided (it makes my Chrome crash).
But thank you for confirming the issue comes from MUI anyway.
Hence I'll close this issue. Feel free to open a new issue on MUI's repo to request for a fix.

As a workaround, you can still rely on the old trick consisting in forcing a re-render by changing the `key` prop.
I have implemented this solution in a [stackblitz](https://stackblitz.com/edit/github-yuwnpr-xryvnm?file=src%2Fposts%2FPostEdit.tsx) forked from yours, and it seems to work fine.

```tsx
  const [count, setCount] = React.useState(1);

  const onClick = () => {
    const newOptions = options.map((v) => ({ ...v, count: v.count + 1 }));
    setOptions(newOptions);
    setCount(count + 1);
  };

          <AutocompleteArrayInput
            // ...
            key={count}
          />
```

**Comment by panfiva** [comment]
@slax57 

I tried using the old trick before I submitted the bug request; however AutocompleteArrayInput was flickering on re-renders in my app (not sure why). I guess I will try using setValue approach to see if that fixes flickering in UI.

FYI, The code from MUI sandbox was as follows

```js
import * as React from "react";
import Autocomplete from "@mui/material/Autocomplete";
import TextField from "@mui/material/TextField";
import Button from "@mui/material/Button";

/**
 * how you used the components
 */
export default function App() {
  const [options, setOptions] = React.useState([
    { label: "movies", id: 1, count: 1 },
    { label: "games", id: 2, count: 1 },
  ]);

  const onClick = () => {
    const newOptions = options.map((v) => ({ ...v, count: v.count + 1 }));
    setOptions(newOptions);
  };

  return (
    <>
      <Autocomplete
        multiple
        id="tags-standard"
        options={options}
        getOptionLabel={(option) => `${option.label} - ${option.count}`}
        isOptionEqualToValue={(option, value) => option.label === value.label}
        renderInput={(params) => (
          <TextField
            {...params}
            variant="standard"
            label="Multiple values"
            placeholder="Favorites"
          />
        )}
      />
      <Button onClick={onClick}>Increase count</Button>
    </>
  );
}

```
